Crypto: Add support for AES accelerator in stm32 crypto driver

str4t0m opened the pull request:
This PR adds hw crypto support for stm32 devices with a hardware AES accelerator.

The existing stm32 driver wraps the STM32Cube HAL functions, and already supports devices with the "CRYP" accelerator.
Therefore only minimal changes are necessary to additionally support devices with the "AES" accelerator.

Additionally to the necessary driver changes, definitions for G0, L5, and WL Series are added.
The crypto sample test configuration is changed such that instead of only building for the mikroe_mini_m4 board, a filter on enabled compatible devices is used, and tests are run in CI.

str4t0m committed:
dts: stm32l5: introduce aes peripheral

This commit adds a dt node for the hw aes accelerator of the
stm32l5 series.

The stm32l562Xe.dtsi file now includes the newly created
stm32l562Xe.dtsi with the additional aes definitions instead of the
stm32l552Xe.dtsi file.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
boards: stm32l562e_dk enable aes peripheral

This commit enables the hw aes accelerator on the stm32l562_dk.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
dts: stm32g0: add definitions for the aes peripheral

This commit adds dt definitions for the hw aes accelerator
on stm32g0 Series.
The accelertor is available on the following socs:
stm32g041, stm32g061, stm32g081, and stm32g0c1.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
dts: stm32wl: add definitions for the aes peripheral

This commit adds a dt node for the hw aes accelerator of the wl series.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
drivers/crypto: stm32: add support for socs with AES IP

This commit adds crypto support for several series which have
the same/very similar or same AES IP.
This includes G0, G4, L5, WL.
WB is also very similar but, expects the app to load the key via CKS.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
samples: drivers: crypto: add support for stm32 with aes peripheral

This commit adds support for stm32 socs with aes peripheral in the
crypto sample.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
boards: nucleo_wl55jc and lora_e5_dev_board enable aes

This commit enables aes support for nucleo_wl55jc and
lora_e5_dev_board.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
samples: drivers.crypto.stm32 test: use filter instead of allow list

This commit uses filter instead of allow list for the
sample.drivers.crypto.stm32 test case.
As a result the ci test is not only built on mikroe_mini_m4_for_stm32,
but also on lora_e5_dev_board, nucleo_wl55jc and stm32l562e_dk.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
samples: crypto: change config to run test sample.drivers.crypto.stm32

The test sample.drivers.crypto.stm32 was built only for the stm32 crypto
driver (build_only: true).
This commit changes the configuration to run the test and adds regex for
the supported modes.
Additionally integration_platform native_posix is removed from the
crypto.stm32 test configuration, such that stm32 targets built in CI.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

erwango commented on drivers/crypto/crypto_stm32.c:
```diff
@@ -408,8 +422,13 @@ static int crypto_stm32_session_free(const struct device *dev,
 		k_sem_give(&data->session_sem);
 		return -EIO;
 	}
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_cryp)
 	__HAL_RCC_CRYP_FORCE_RESET();
 	__HAL_RCC_CRYP_RELEASE_RESET();
+#elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32_aes)
+	__HAL_RCC_AES_FORCE_RESET();
+	__HAL_RCC_AES_RELEASE_RESET();
+#endif
```
Nit: Provide abstraction macros here and for `xxx_TypeDef` to group the #idefery in a single place.

str4t0m replied:
I have now declared a few macros at the beginning of the file, please let me know if you had something different in mind.
Additionally I renamed `STM32_CRYP_KEYSIZE_192B_SUPPORT` to `STM32_CRYPTO_KEYSIZE_192B_SUPPORT`.

erwango commented on dts/arm/st/wl/stm32wl.dtsi:
```diff
@@ -382,6 +382,15 @@
 		};
 
+		aes: aes@58001800 {
+			compatible = "st,stm32-aes";
```
We miss a binding for this compatible

str4t0m replied:
Indeed, added it now as the first commit.

erwango commented on samples/drivers/crypto/src/main.c:
```diff
@@ -21,8 +21,10 @@ LOG_MODULE_REGISTER(main);
 #elif CONFIG_CRYPTO_MBEDTLS_SHIM
 #define CRYPTO_DRV_NAME CONFIG_CRYPTO_MBEDTLS_SHIM_DRV_NAME
-#elif CONFIG_CRYPTO_STM32
+#elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32_cryp) && CONFIG_CRYPTO_STM32
```
I'm not sure `CONFIG_CRYPTO_STM32` is still required

str4t0m replied:
It is not required any more because timycrypt and mbedtls are checked before. I have removed it now.

erwango commented on samples/drivers/crypto/src/main.c:
```diff
@@ -21,8 +21,10 @@ LOG_MODULE_REGISTER(main);
-#elif CONFIG_CRYPTO_STM32
+#elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32_cryp) && CONFIG_CRYPTO_STM32
 #define CRYPTO_DRV_NAME DT_LABEL(DT_INST(0, st_stm32_cryp))
+#elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32_aes) && CONFIG_CRYPTO_STM32
```
ditto

erwango commented on boards/arm/lora_e5_dev_board/lora_e5_dev_board.dts:
```diff
@@ -149,6 +149,10 @@
 	status = "okay";
 };
 
+&aes {
```
Would be nice to update board doc

erwango commented on boards/arm/nucleo_wl55jc/nucleo_wl55jc.dts:
```diff
@@ -131,6 +131,10 @@
+&aes {
+	status = "okay";
+};
+
```
Would be nice to update board doc

erwango commented on boards/arm/stm32l562e_dk/stm32l562e_dk.dts:
```diff
@@ -48,3 +48,7 @@
 &iwdg {
 	status = "okay";
 };
+
```
Would be nice to update board doc

str4t0m replied:
I have updated the board docs in a single commit.
As some other enabled peripherals were not listed as supported, I have added those as well.
Additionally I have ordered the features alphabetically, and added some links to the board configuration files.

erwango commented on dts/arm/st/g0/stm32g0_crypt.dtsi:
```diff
@@ -10,6 +10,15 @@
 	};
 
 	soc {
+		aes: aes@40026000 {
```
Would be nice to enable on a board, so it could be at least built in CI

str4t0m replied:
The only official ST boards with HW AES support are STM32G081B-EVAL and STM32G0C1E-EV.
I assume neither are widely distributed.
But regular CI builds and eventually test runs would indeed be very helpful.
Would it be fine if we add such a board later in a extra pr? Would anybody of you have access to such a board to run tests, or would it be for build tests only?

erwango reviewed (commented):
Thanks for this PR, few comments but otherwise this could go in a PR

str4t0m committed:
dts: bindings: add ST STM32 AES accelerator

This commit adds bindings for the STM32 AES accelerator.
Common properties of st,stm32-cryp are moved into
st,stm32-crypto-common.yaml.

The accelerator supports ECB, CBC, CTR, GCM, GMAC, and CCM
chaining modes.
128 bit bit data blocks with cipher key lengths of 128 and 256 bit are
supported.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
doc: stm32wl and stm32l5 boards: update supported features

This commit updates the supported features of stm32l5 and stm32wl boards
to match the devicetree configuration.
- add link to board defconfig and board.dts
- sort supported features table alphabethically by interface.

Additionally CRYP peripheral is added as supported feature to the
mikroe_mini_m4_for_stm32 board documentation.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

ABOSTM reviewed (approved):
LGTM

mbolivar-nordic commented on drivers/crypto/crypto_stm32.c:
```diff
@@ -21,12 +19,34 @@
 #include <logging/log.h>
 LOG_MODULE_REGISTER(crypto_stm32);
 
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_cryp)
+#define DT_DRV_COMPAT st_stm32_cryp
+#elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32_aes)
+#define DT_DRV_COMPAT st_stm32_aes
```
Can both be enabled at the same time on the same SoC? If so, can you please make the driver multi-instance and support both cases?

Note you can do something like this to handle that within a single driver if that makes sense:

```c
#define DT_DRV_COMPAT st_stm32_cryp
DT_INST_FOREACH_STATUS_OKAY(DEFINE_CRYP)
#undef DT_DRV_COMPAT
#define DT_DRV_COMPAT st_stm32_aes
DT_INST_FOREACH_STATUS_OKAY(DEFINE_AES)
```

If the device is a singleton and no SoC contains both, then this PR LGTM too.

str4t0m replied:
Thanks for this suggestions!
In this case no SoC contains both, but only a single instance of either AES or CRYP.

cfriedt commented:
@erwango - is this ready to be merged?

str4t0m commented on drivers/crypto/Kconfig.stm32:
```diff
@@ -3,8 +3,12 @@
+
 menuconfig CRYPTO_STM32
 	bool "STM32 Cryptographic Accelerator driver"
+	default $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_AES)) || $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_CRYP))
 	depends on SOC_FAMILY_STM32
```
Probably it would be better to move the dt_compat_enabled from `default` to `depends on`:
```suggestion
	depends on $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_AES)) || $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_CRYP))`
```
Otherwise the driver will also be built if the TinyCrypt or mbedTLS stubs are used and needs to be explicitly disabled.
While this doesn't cause any conflicts it would waste flash/ram. Please let me know which one you prefer.

cfriedt replied:
@erwango - please comment here

cfriedt replied:
@str4t0m - I'm going to approve and merge. As it stands, this is the worse scenario, if I understand correctly. Reducing flash / ram could be considered an optimization and can be deferred. If you would like to make this change, please create a separate issue, reference this PR, and provide some numbers for ram / flash footprints with and without the change.

Thanks again for getting this in!
